src/buffer/out/sgrStack.hpp

```diff
@@ -0,0 +1,238 @@
+            //  CSI Ps ; Ps # {
+            //            Push video attributes onto stack (XTPUSHSGR), xterm.  The
+            //            optional parameters correspond to the SGR encoding for video
+            //            attributes, except for colors (which do not have a unique SGR
+            //            code):
+            //              Ps = 1  -> Bold.
+            //              Ps = 2  -> Faint.
+            //              Ps = 3  -> Italicized.
+            //              Ps = 4  -> Underlined.
+            //              Ps = 5  -> Blink.
+            //              Ps = 7  -> Inverse.
+            //              Ps = 8  -> Invisible.
+            //              Ps = 9  -> Crossed-out characters.
+            //              Ps = 1 0  -> Foreground color.
+            //              Ps = 1 1  -> Background color.
+            //              Ps = 2 1  -> Doubly-underlined.
+            //
+            //  (some closing braces for people with editors that get thrown off without them: }})
+            //
+            // Attributes that are not currently supported are simply ignored.
+
+            if (_GraphicsOptionToFlag(DispatchTypes::GraphicsOptions::BoldBright) & validParts)
+            {
+                if (savedAttribute.IsBold())
+                {
+                    result.Embolden();
+                }
+                else
+                {
+                    result.Debolden();
+                }
+            }
+
+            if (_GraphicsOptionToFlag(DispatchTypes::GraphicsOptions::Underline) & validParts)
+            {
+                if (savedAttribute.IsUnderline())
+                {
+                    result.EnableUnderline();
+                }
+                else
+                {
+                    result.DisableUnderline();
+                }
+            }
+
+            if (_GraphicsOptionToFlag(DispatchTypes::GraphicsOptions::Negative) & validParts)
+            {
+                if (savedAttribute.IsReverseVideo())
+                {
+                    if (!result.IsReverseVideo())
+                    {
+                        result.Invert();
+                    }
+                }
+                else
+                {
+                    if (result.IsReverseVideo())
+                    {
+                        result.Invert();
+                    }
+                }
+            }
+
+            if (_GraphicsOptionToFlag(DispatchTypes::GraphicsOptions::SaveForegroundColor) & validParts)
+            {
+                result.SetForegroundFrom(savedAttribute);
+            }
+
+            if (_GraphicsOptionToFlag(DispatchTypes::GraphicsOptions::SaveBackgroundColor) & validParts)
+            {
+                result.SetBackgroundFrom(savedAttribute);
+            }
+
+            return result;
+        }
+
+        int _numSgrPushes; // used as an index into the following arrays
+        TextAttribute _storedSgrAttributes[c_MaxStoredSgrPushes];
```

Comment:
Can these be std::arrays? Then you could use .at on them and eliminate the warning suppression above.

Reply:
Another option would be a `std::stack` of a simple `struct` type containing the `TextAttribute` and `uint32_t` mask (or a `std::bitset`). That's possibly less efficient, but at least it's standard functionality, and it keeps the stack management separate from the (de)serialization complexity, which could either be part of the `struct`, or the `AdaptDispatch` code that pushes and pops the attributes.

Reply:
I batched the `TextAttribute` and `bitset` into a struct, and store those in a single `std::array`. I didn't go with `std::stack`, because I've implemented a "ring stack". (lot of comments about why in sgrStack.hpp)